File: vite-project/vite-project/src/models/Question.ts
import Item from "./Item";

/**
 * Question class
 * 
 * Represents a question in the game.
 * 
 * @author Zach Sanchez
 * @version 1.0
*/

export default class Question{
    /**
     * Static string for incorrect text
    */
    public static INCORRECT_TEXT = "INCORRECT_ANSWER";

    /**
     * The question text, with a "!" to signify that it will force initialization if not set
     */
    private myQuestionText!: string;

    /**
     * The correct answer, with a "!" to signify that it will force initialization if not set
     */
    private myCorrectAnswer!: string;

    /**
     * The choices for the question, with a "!" to signify that it will force initialization if not set
     */
    private myChoices!: string[];

    constructor(theQuestionText: string, theCorrectAnswer: string, theChoices: string[]) {
        this.myQuestionText = this.setMyQuestionText(theQuestionText);
        this.myCorrectAnswer = this.setMyCorrectAnswer(theCorrectAnswer);
        this.myChoices = this.setMyChoices(theChoices);
    }


    /**
     * Applies an item to the question
     * 
     * @param {Item} theItem - The item to apply
     * @returns {void}
    */
    public applyItem(theItem: Item): void {
        let itemType = theItem.getItemType();
        if(itemType === "50-50"){
            this.applyFiftyFifty();
        } else if(itemType === "Phone-a-Friend"){
            this.applyPhoneAFriend();
        }
    }

    /**
     * Returns a random wrong answer
     * 
     * @returns {string} - A random wrong answer
    */
    private wrongAnswer(): string {
        let randomIndex = Math.floor(Math.random() * this.myChoices.length);
        let wrongAnswers = this.myChoices.filter(choice => choice !== this.myCorrectAnswer);

        return wrongAnswers[randomIndex];
    }

    /**
     * Applies the Phone-a-Friend item to the question
     * There is a 10% chance that the friend will give the wrong answer
     * 
     * @returns {void}
    */
    private applyPhoneAFriend(): void {
        let chance = Math.random();
        if(chance < 0.1) {
            alert( this.wrongAnswer());
        } else {
            alert(this.myCorrectAnswer);
        }
    }

    /**
     * Applies the Fifty-Fifty item to the question
     * Eliminates two wrong answers
     * 
     * @returns {void}
    */
    private applyFiftyFifty(): void {
        if(this.myChoices.includes(Question.INCORRECT_TEXT)){
            console.log("Fifty fifty can only be applied on the first turn.")
            return
        }
        let elim = 0

        while(elim < 2) {
            let wrongAnswer = this.wrongAnswer();
            if(wrongAnswer !== Question.INCORRECT_TEXT){
                this.myChoices.splice(this.myChoices.indexOf(wrongAnswer), 1);
                elim++;
            }
        }
        console.log("Fifty fifty eliminated 2 choices")
    }

    /**
     * Invalidates a choice
     * 
     * @param {number} theInvalidChoice - The index of the choice to invalidate
     * @returns {void}
    */
    private invalidateChoice(theInvalidChoice:number): void {
        this.myChoices[theInvalidChoice] = Question.INCORRECT_TEXT;
    }

    /**
     * Asks the question and returns whether the answer is correct
     * 
     * @returns {boolean} - Whether the answer is correct
    */
    public askQuestion():boolean{
        this.printQuestion();
        this.printChoices();

        let answer = prompt("Enter the number of your answer: ");
        if(answer === null){
            console.log("Not a valid answer. Please enter a number");
            return false;
        }
        let answerNumber = parseInt(answer)-1;
        let theChosenAnswer = this.myChoices[answerNumber];
        let isCorrect = theChosenAnswer === this.myCorrectAnswer;
        if(!isCorrect){
            this.invalidateChoice(answerNumber);
        }

        this.printResult(isCorrect);
        return isCorrect;
    }
    
    /**
     * Gets the question text
     * 
     * @returns {string} - The question text
    */
    public getMyQuestionText(): string {
        return this.myQuestionText;
    }

    /**
     * Gets the correct answer
     * 
     * @returns {string} - The correct answer
    */
    public getMyCorrectAnswer(): string {
        return this.myCorrectAnswer;
    } 

    /**
     * Gets the choices for the question
     * 
     * @returns {string[]} - The choices for the question
     */
    public getMyChoices(): string[] {
        return this.myChoices;
    }   

    /**
     * Sets the questions text
     * 
     * @param {string} theQuestionText - The text to set as the question text
     * @returns {string} - The question text
     */
    public setMyQuestionText(theQuestionText: string): string {
        if(theQuestionText.length === 0) {
            throw new Error(Question.INCORRECT_TEXT);
        }
       return theQuestionText;
    }

    /**
     * Sets the correct answer
     * 
     * @param {string} theCorrectAnswer - The correct answer to set
     * @returns {string} - The correct answer
     */
    public setMyCorrectAnswer(theCorrectAnswer: string): string {
        if(theCorrectAnswer.length === 0) {
            throw new Error(Question.INCORRECT_TEXT);
        }
        return theCorrectAnswer;
    }


    /**
     * Sets the choices for the question
     * 
     * @param {string[]} theChoices - The choices to set
     * @returns {string[]} - The choices for the question
     */
    public setMyChoices(theChoices: string[]): string[] {
        if(theChoices.length === 0) {
            throw new Error(Question.INCORRECT_TEXT);
        }
        return theChoices;
    }


    /**
     * Prints the question
     * 
     * @returns {void}
    */
    private printQuestion(): void {
        console.log(this.myQuestionText);
    }

    /**
     * Prints the choices
     * 
     * @returns {void}
    */
    private printChoices(): void {
        let choiceNumber = 1;
        this.myChoices.forEach(choice => {
            console.log(choiceNumber + ". " + choice);
            choiceNumber++;
        });
    }

    /**
     * prints the result of the question
     * 
     * @param {boolean} result - Whether the answer is correct
     */
    private printResult(result: boolean): void {
        if(result) {
            console.log("Correct!");
        } else {
            console.log("Incorrect!");
        }
    }
    
    /**
     * Returns the question text as a string
     * 
     * @returns {string} - The question text
     */
    public toString(): string {
        return this.myQuestionText;
    }    
}

File: vite-project/src/model/Question.java
package model;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.Scanner;

/**
 * The Question class represents a trivia question used in the Trivia Maze game.
 * It contains the question text, a list of possible choices, and the correct answer.
 * This class also handles the logic for asking the question and checking if the 
 * player's answer is correct.
 * 
 * @author Zach Sanchez (zachs00)
 * @version October 26th, 2024
 */
public class Question {
    
    /**
     * A static string used to mark an answer as incorrect.
     */
    private static String INCORRECT_TEXT = "INCORRECT_ANSWER";
    
    /**
     * The text of the question.
     */
    private final String myQuestionText;
    
    /**
     * The correct answer for this question.
     */
    private final String myCorrectAnswer;
    
    /**
     * A list of possible answer choices for this question.
     */
    private List<String> myChoices;

    /**
     * Constructs a new Question object with the provided question text, 
     * correct answer, and list of answer choices.
     * 
     * @param theQuestionText The text of the question.
     * @param theCorrectAnswer The correct answer for the question.
     * @param theChoices A list of possible answer choices.
     */
    public Question(String theQuestionText, String theCorrectAnswer, List<String> theChoices) {
        this.myQuestionText = setMyQuestionText(theQuestionText);
        this.myCorrectAnswer = setMyCorrectAnswer(theCorrectAnswer);
        this.myChoices = setMyChoices(theChoices);
    }
    
    /**
     * Applies the item's effect to the current question.
     * 
     * If the item is "Phone-a-Friend", it will reveal the correct answer.
     * If the item is "50-50", it will randomly eliminate two incorrect answers from myChoices.
     * 
     * @param theItem the item to be applied, such as "Phone-a-Friend" or "50-50".
     */
    public void applyItem(Item theItem) {
        String itemType = theItem.getItemType();
        if (itemType.equals("Phone-a-Friend")) {
            this.applyPhoneAFriend();
        } else if (itemType.equals("50-50")) {
            this.applyFiftyFifty();
        }
    }

    /**
     * Phone-a-Friend power-up.
     * 
     * Simulates a 'phone a friend' life-line. There is a 10% chance that the friend 
     * will give the wrong answer, otherwise they give the correct answer.
     */
    private void applyPhoneAFriend() {
        Random random = new Random();
        int chance = random.nextInt(100);
        
        if (chance < 10) {
            String wrongAnswer = getRandomWrongAnswer();
            System.out.println("Phone-a-Friend: says '" + wrongAnswer + "' is the correct answer.");
        } else {
            System.out.println("Phone-a-Friend: says '" + this.getMyCorrectAnswer() + "' is the correct answer.");
        }
    }
    
    /**
     * Chooses a random incorrect answer.
     * 
     * @return a random incorrect answer.
     */
    private String getRandomWrongAnswer() {
        Random random = new Random();
        List<String> wrongAnswers = new ArrayList<>();
        
        for (String choice : myChoices) {
            if (!choice.equals(myCorrectAnswer) && !choice.equals(INCORRECT_TEXT)) {
                wrongAnswers.add(choice);
            }
        }
        
        return wrongAnswers.get(random.nextInt(wrongAnswers.size()));
    }

    /**
     * 50-50 power-up.
     * 
     * Randomly eliminates two incorrect answers from the current choices, ensuring
     * that the correct answer is never removed. If any incorrect choices have
     * already been eliminated (marked by "INCORRECT_TEXT"), this power-up cannot be used.
     */
    private void applyFiftyFifty() {

        if (myChoices.contains(INCORRECT_TEXT)) {
            System.out.println("This can only be used on the first turn.");
            return;
        }

        Random rand = new Random();
        int eliminated = 0;
        int correctAnswerIndex = myChoices.indexOf(myCorrectAnswer);

        while (eliminated < 2) {
            int idx = rand.nextInt(myChoices.size());
            if (idx != correctAnswerIndex && !myChoices.get(idx).equals(INCORRECT_TEXT)) {
                myChoices.set(idx, INCORRECT_TEXT);
                eliminated++;
            }
        }

        System.out.println("Fifty-fifty applied. Two incorrect answers have been removed.");
    }

    
    /**
     * Marks a specific choice as incorrect by replacing it with a static string.
     * 
     * @param theInvalidChoiceIdx The index of the incorrect choice to invalidate.
     */
    private void invalidateChoice(int theInvalidChoiceIdx) {
        this.myChoices.set(theInvalidChoiceIdx, INCORRECT_TEXT);
    }
    
    /**
     * Asks the player the question, collects their answer, checks if it's correct,
     * and returns the result as a boolean.
     * 
     * @return true if the player's answer is correct, false if it is incorrect.
     */
    public boolean askQuestion() {
        this.printQuestion();
        this.printChoices();
        System.out.println("Enter an int to select a choice: ");
        
        Scanner scanny = new Scanner(System.in);
        int theChosenAnswerIdx = scanny.nextInt() - 1;
        String theChosenAnswer = myChoices.get(theChosenAnswerIdx);
        boolean isCorrect = theChosenAnswer.equals(this.myCorrectAnswer);
        if (!isCorrect) {
            this.invalidateChoice(theChosenAnswerIdx);
        }
        
        this.printResult(isCorrect);
        return isCorrect;
    }
    
    /**
     * Returns the text of the question.
     * 
     * @return The question text.
     */
    public String getMyQuestionText() {
        return myQuestionText;
    }

    /**
     * Returns the correct answer for the question.
     * 
     * @return The correct answer.
     */
    public String getMyCorrectAnswer() {
        return myCorrectAnswer;
    }

    /**
     * Returns the list of possible answer choices for the question.
     * 
     * @return A list of answer choices.
     */
    public List<String> getMyChoices() {
        return myChoices;
    }
    
    /**
     * Validates and sets the question text. If the text is null or empty, 
     * an exception is thrown.
     * 
     * @param theQuestionText The text of the question.
     * @return The validated question text.
     * @throws IllegalArgumentException if the text is null or empty.
     */
    private String setMyQuestionText(String theQuestionText) {
        if (theQuestionText != null && !theQuestionText.isEmpty()) {
            return theQuestionText;
        }
        throw new IllegalArgumentException("Invalid Value: Question text was blank/null.");
    }
    
    /**
     * Validates and sets the correct answer. If the answer is null or empty,
     * an exception is thrown.
     * 
     * @param theQuestionAnswer The correct answer.
     * @return The validated correct answer.
     * @throws IllegalArgumentException if the answer is null or empty.
     */
    private String setMyCorrectAnswer(String theQuestionAnswer) {
        if (theQuestionAnswer != null && !theQuestionAnswer.isEmpty()) {
            return theQuestionAnswer;
        }
        throw new IllegalArgumentException("Invalid Value: Answer text was blank/null.");
    }
    
    /**
     * Validates and sets the list of answer choices. If the list is null or empty,
     * an exception is thrown.
     * 
     * @param theChoices A list of possible answer choices.
     * @return The validated list of answer choices.
     * @throws IllegalArgumentException if the list is null or empty.
     */
    private List<String> setMyChoices(List<String> theChoices) {
        if (theChoices != null && !theChoices.isEmpty()) {
            return theChoices;
        }
        throw new IllegalArgumentException("Invalid Value: Choices list was blank/null.");
    }
    
    /**
     * Prints the text of the question.
     */
    private void printQuestion() {
        System.out.println(this.getMyQuestionText());
    }
    
    /**
     * Prints the list of possible answer choices, each with a number.
     */
    private void printChoices() {
        List<String> theChoices = this.getMyChoices();
        int count = 1;
        for (String choice : theChoices) {
            System.out.println(count + ": " + choice);
            count++;
        }
    }
    
    /**
     * Prints the result of the player's answer: "Correct!" if they were right, 
     * or "Incorrect!" if they were wrong.
     * 
     * @param result true if the answer was correct, false otherwise.
     */
    private void printResult(boolean result) {
        if (result) {
            System.out.println("Correct!");
        } else {
            System.out.println("Incorrect!");
        }
    }

    
    public String toString() {
    	return this.getMyQuestionText();
    }

}
